src/Adapter/Grid/Data/Factory/CartGridDataFactory.php

```diff
@@ -0,0 +1,138 @@
+ */
+final class CartGridDataFactory implements GridDataFactoryInterface
+{
+    /**
+     * @var GridDataFactoryInterface
+     */
+    private $doctrineDataFactory;
+
+    /**
+     * @var TranslatorInterface
+     */
+    private $translator;
+
+    /**
+     * @var LocaleInterface
+     */
+    private $locale;
+
+    /**
+     * @param GridDataFactoryInterface $doctrineDataFactory
+     * @param TranslatorInterface $translator
+     * @param LocaleInterface $locale
+     */
+    public function __construct(
+        GridDataFactoryInterface $doctrineDataFactory,
+        TranslatorInterface $translator,
+        LocaleInterface $locale
+    ) {
+        $this->doctrineDataFactory = $doctrineDataFactory;
+        $this->translator = $translator;
+        $this->locale = $locale;
+    }
+
+    /**
+     * {@inheritdoc}
+     */
+    public function getData(SearchCriteriaInterface $searchCriteria)
+    {
+        $data = $this->doctrineDataFactory->getData($searchCriteria);
+
+        $records = $this->applyModifications($data->getRecords());
+
+        return new GridData($records, $data->getRecordsTotal(), $data->getQuery());
+    }
+
+    /**
+     * @param RecordCollectionInterface $records
+     *
+     * @return RecordCollection
+     */
+    private function applyModifications(RecordCollectionInterface $records)
+    {
+        $modifiedRecords = [];
+
+        foreach ($records as $record) {
+            switch ($record['status']) {
+                case 'abandoned_cart':
+                    $record['status'] = $this->translator->trans(
+                        'Abandoned cart',
+                        [],
+                        'Admin.Orderscustomers.Feature'
+                    );
+                    break;
+                case 'not_ordered':
+                    $record['status'] = $this->translator->trans('Not placed', [], 'Admin.Orderscustomers.Feature');
+                    break;
+            }
+
+            if (empty($record['carrier_name'])) {
+                $record['carrier_name'] = '--';
+            }
+
+            if (empty($record['customer_name'])) {
+                $record['customer_name'] = '--';
+            }
+
+            $record['online'] = $record['id_guest'] ?
+                $this->translator->trans('Yes', [], 'Admin.Global') :
+                $this->translator->trans('No', [], 'Admin.Global');
+
+            $context = Context::getContext();
+            $context->currentLocale = $this->locale;
+            $context->cart = new Cart($record['id_cart']);
+            $context->currency = new Currency((int) $context->cart->id_currency);
+            $context->customer = new Customer((int) $context->cart->id_customer);
+
+            $record['cart_total'] = Cart::getTotalCart($context->cart->id, true, Cart::BOTH_WITHOUT_SHIPPING);
+            $record['is_order_placed'] = $context->cart->orderExists();
+            $record['shop_name'] = $context->shop->getShops()[$context->cart->id_shop]['name'];
```

Comment:
May be check that `$context->cart->id_shop` isset in the array

Reply:
Refactored like this

```
            if (!isset($context->cart->id_shop)) {
                throw new CartException('cart shop id is not set');
            }
            $record['shop_name'] = $context->shop->getShops()[$context->cart->id_shop]['name'];
```